### ISSUE TITLE
GCSFuse process crashing without logging anything

### ISSUE THREAD

**Issue by davidkrikava** [GoogleCloudPlatform/gcsfuse#1603]
**System:**
Google VM instance (E2 machine type)
CentOS Linux release 7.9.2009
GCFuse version: 1.4.0 (issue was on 1.3.0 also)
Bucket in use: regional (mounted from two VMs from different zones within that region)

**mount options:**
`bucketname /mount/path _netdev,noatime,rw,stat_cache_capacity=81920,max_retry_sleep=0,stat_cache_ttl=5m0s,type_cache_ttl=30m0s,client_protocol=http2,max_conns_per_host=200,allow_other,file_mode=0775,dir_mode=0775,uid=XXX,gid=XXX,log_file=/var/log/gcsfuse.log,debug_fuse,debug_gcs,debug_mutex`

**VM resources:**
during crashes were nowhere near exhausting VMs resources
RAM: ~30% used
cpu: ~10% used

**Sizing, req rates:**
bucket size ~ 5,5T
object count ~ 1,3M
req rate during crashes ~ 100/s
(above info taken from google metrics explorer from the bucket)

**Issue description:**
We have two vms (01 and 02) that have regional bucket mounted via gcsfuse. In that bucket we have two folders 01 and 02 (and then a subtree of folders/files), machines write into their folders respectively, but read from both.
Both machines also cleanup their respective folders every morning and we think this cleanup is whats causing the crash as the time of the crashes correlates with the high req count.

Cleanup command (yes this is not optimal, we should use bucket lifecycle instead, but we switched to cron/find script long time ago for some reason):
`find /mountpoint/01/ -mtime +20 -type f | while read path; do rm "$path"; done`
First occurrence: 25. Dec 2023 (version 1.3.0, continue happening with 1.4.0 version also)
**Just the find part is enough to cause gcsfuse crash.**
What actually happens: gcsfuse process crashes, no longer running, can be fixed by first unmounting and then mounting again

**Logs:**
As I posted in "mount options" - we are using several debug options, yet when the crash occurs there is nothing logged.
Log file ends with last successful transaction.
Same with syslog, there is nothing logged, only first failed access:
`Jan 10 10:56:37 server-01 collectd[1440]: statvfs(/XXX) failed: Transport endpoint is not connected`
Nothing can be found in dmesg either.

**To sum up:**
I am not aware of reaching any quotas (maybe there are some I am not aware ? - tho i would expect to have errors logged instead of process cashing in such case), this process has been running without changes (and previously until that date without issues) for the past 7 months.
What can we do, to try and figure out the root cause of these crashes, since there is nothing logged anywhere?
Thank you for your help

**Comment by raj-prince** [comment]
Thanks @davidkrikava for reaching out and detailed description!

I am looking into this, in the mean time I need one clarification. Do you see this issue after every 20hrs `(-mtime +20)`? In other words, is it 100% reproducible?

**Comment by davidkrikava** [comment]
Unfortunately it is not, so far we had crashes
25. Dec, 27. Dec, 5. Jan, 9. Jan, 10. Jan - all during the higher req rate caused by the find command (but the command runs every day)
Then second one 10. Jan, when i tried to run the find command manually.
<img width="858" alt="image" src="https://github.com/GoogleCloudPlatform/gcsfuse/assets/156192824/0f660b01-266f-4281-80ee-7e8ded1b05c5">
Here one can see the diff, when the command runs vs normal traffic (last 7 days)
Thank you for swift response.

**Comment by raj-prince** [comment]
Hi @davidkrikava, thanks for the patience!

GCSFuse doesn't log when killed externally, but it does log stack traces for crashes within the program in most of the cases. 

GCSFuse logs the reason for crash and stack-trace if it is mounted in foreground mode (use `--foreground` for this). So, could you please try mounting GCSFuse in foreground mode?

When it is mounted in background mode (by default), it logs the crashes info other than crash in [jacobsa/fuse](https://github.com/jacobsa/fuse) (one of the core dependencies of gcsfuse). 

In the mean, time I tried reproducing this but couldn't succeeded yet. If you could provide the minimize steps to reproduce this issue, that would be good for us.

Also, you can consider option to clean the object using [storage_lifecyle_delete](https://cloud.google.com/storage/docs/lifecycle#delete) and see if this comes with that too.

 Regards,
 Prince Kumar.

**Comment by davidkrikava** [comment]
Hi,
well unfortunately I cannot provide 100% working steps to reproduce it, but so far, I was able to bring it down once myself like this:
(also please consider the sizing from original post)
Try mounting regional bucket from two zones (two VMs)
Run find command from both VMs simultaneously (with some deletes)
"Hope" for crash.

At this point I remounted with `--foreground` as you suggested and trying to bring it down myself to see if there will be something logged.
Running the gcfuse like this: `gcsfuse --foreground --stat-cache-ttl 5m0s --type-cache-ttl 30m0s -o allow_other --dir-mode 0775 --uid XXX --debug_mutex -o rw -o noatime --gid XXX --debug_fuse --debug_gcs --max-retry-sleep 0 --max-conns-per-host 200 --log-file /var/log/gcsfuse.log --stat-cache-capacity 81920 --client-protocol http2 --file-mode 0775 bucket-name /our/path &`
Thanks

**Comment by davidkrikava** [comment]
Hi,
today i finally got lucky and captured an output with whole stacktrace
so I ran the command as i described:

`gcsfuse --foreground --stat-cache-ttl 5m0s --type-cache-ttl 30m0s -o allow_other --dir-mode 0775 --uid XXX --debug_mutex -o rw -o noatime --gid XXX --debug_fuse --debug_gcs --max-retry-sleep 0 --max-conns-per-host 200 --log-file /var/log/gcsfuse.log --stat-cache-capacity 81920 --client-protocol http2 --file-mode 0775 bucket-name /our/path`

leaving it in a screen, so if the process crashed, I would get the output in the command line - and then i tried running those cleanup commands (finds with delete)
The output i got was -

`fatal error: concurrent map iteration and map write`

whole trace i captured:
[gcsfuse_crash_stacktrace.txt](https://github.com/GoogleCloudPlatform/gcsfuse/files/13939338/gcsfuse_crash_stacktrace.txt)

**Comment by sethiay** [comment]
@davidkrikava Thank you for sharing the crash logs. That was helpful and we are able to reproduce the issue on our side. We are working on next steps/fix and will keep you updated on the same.

**Comment by ashmeenkaur** [comment]
Hi @davidkrikava, 
just wanted to follow up - were you able to use the mitigation we suggested in https://github.com/GoogleCloudPlatform/gcsfuse/issues/1603#issuecomment-1895689084?

**Comment by davidkrikava** [comment]
Yes, I thought as well that the delete was the issue, but it seems that the excessive listing by the find combined with regular usage (we have nginx root serving files from there - creating and listing) is the issue.

We are moving towards the bucket lifecycle.

Yes than we will definitely update once the fix is released, as IMO the listing greatly increased the chances of a crash, but since we do create files there and read them, this can theoretically happen at anytime now.

I now consider my participation in this case finished, but do contact me, if you need anything else in the future.
Thank you

**Comment by raj-prince** [comment]
Thanks @davidkrikava!

Lowering the priority of this issue, but keeping this open until we fix the crash.

**Comment by ashmeenkaur** [comment]
Hi @davidkrikava, 

Thanks for reporting this issue! The fix has been merged into master and will be included in the March 2024 release.

Thanks, 
Ashmeen

**Comment by Tulsishah** [comment]
This issue was resolved in GCSFuse [v2.0.0](https://github.com/GoogleCloudPlatform/gcsfuse/releases/tag/v2.0.0). Please upgrade and reopen the issue if necessary.

Thanks,
Tulsi Shah
